common/scala/src/main/scala/whisk/core/database/s3/S3AttachmentStore.scala

```diff
@@ -0,0 +1,192 @@
+  }
+
+  override protected[core] def readAttachment[T](docId: DocId, name: String, sink: Sink[ByteString, Future[T]])(
+    implicit transid: TransactionId): Future[T] = {
+    require(name != null, "name undefined")
+    val start =
+      transid.started(
+        this,
+        DATABASE_ATT_GET,
+        s"[ATT_GET] '$prefix' finding attachment '$name' of document 'id: $docId'")
+    val (source, _) = client.download(bucket, objectKey(docId, name))
+
+    val f = source.runWith(sink)
+
+    val g = f.transform(
+      { s =>
+        transid
+          .finished(this, start, s"[ATT_GET] '$prefix' completed: found attachment '$name' of document 'id: $docId'")
+        s
+      }, {
+        case s: Throwable if isMissingKeyException(s) =>
+          transid
+            .finished(
+              this,
+              start,
+              s"[ATT_GET] '$prefix', retrieving attachment '$name' of document 'id: $docId'; not found.")
+          NoDocumentException("Not found on 'readAttachment'.")
+        case e => e
+      })
+
+    reportFailure(
+      g,
+      start,
+      failure =>
+        s"[ATT_GET] '$prefix' internal error, name: '$name', doc: 'id: $docId', failure: '${failure.getMessage}'")
+  }
+
+  override protected[core] def deleteAttachments(docId: DocId)(implicit transid: TransactionId): Future[Boolean] = {
+    val start =
+      transid.started(this, DATABASE_ATTS_DELETE, s"[ATT_DELETE] deleting attachments of document 'id: $docId'")
+
+    //S3 provides API to delete multiple objects in single call however alpakka client
+    //currently does not support that and also in current usage 1 docs has at most 1 attachment
+    //so current approach would also involve 2 remote calls
+    val f = client
+      .listBucket(bucket, Some(objectKeyPrefix(docId)))
+      .mapAsync(1)(bc => client.deleteObject(bc.bucketName, bc.key))
+      .runWith(Sink.seq)
+      .map(_ => true)
+
+    f.onSuccess {
+      case _ =>
+        transid.finished(this, start, s"[ATTS_DELETE] completed: deleting attachments of document 'id: $docId'")
+    }
+
+    reportFailure(
+      f,
+      start,
+      failure => s"[ATTS_DELETE] '$prefix' internal error, doc: '$docId', failure: '${failure.getMessage}'")
+  }
+
+  override protected[core] def deleteAttachment(docId: DocId, name: String)(
+    implicit transid: TransactionId): Future[Boolean] = {
+    val start =
+      transid.started(this, DATABASE_ATT_DELETE, s"[ATT_DELETE] deleting attachment '$name' of document 'id: $docId'")
+
+    val f = client
+      .deleteObject(bucket, objectKey(docId, name))
+      .map(_ => true)
+
+    f.onSuccess {
+      case _ =>
+        transid.finished(this, start, s"[ATT_DELETE] completed: deleting attachment '$name' of document 'id: $docId'")
+    }
+
+    reportFailure(
+      f,
+      start,
+      failure => s"[ATT_DELETE] '$prefix' internal error, doc: '$docId', failure: '${failure.getMessage}'")
+  }
+
+  override def shutdown(): Unit = {}
```

Comment:
nothing to shutdown?

Reply:
It uses Akka cached Http connection pool which is bound to ActorSystem lifecycle so need not be explicitly closed. It may be possible to close the pool but [comments here][1] indicates that it may pose problem so not attempting them

[1]: https://github.com/apache/incubator-openwhisk/pull/3870/files#diff-d6387ae1a4ba277912dd09a216b1f92bL127